mlflow/utils/environment.py

```diff
@@ -18,6 +25,91 @@
 _CONDA_ENV_FILE_NAME = "conda.yaml"
 _REQUIREMENTS_FILE_NAME = "requirements.txt"
 _CONSTRAINTS_FILE_NAME = "constraints.txt"
+_PYTHON_ENV_FILE_NAME = "python-env.yaml"
+
+
+class EnvManager(Enum):
+    LOCAL = "local"
+    CONDA = "conda"
+    VIRTUALENV = "virtualenv"
+
+
+_PYTHON_REQUIREMENT_REGEX = re.compile(r"^python=([\d.]+)$")
+
+
+class PythonEnv:
+    def __init__(self, python=None, build_dependencies=None, dependencies=None):
+        self.python = python
+        self.build_dependencies = build_dependencies
+        self.dependencies = dependencies
+
+    def __str__(self):
+        return str(self.to_dict())
+
+    @staticmethod
+    def get_current_python():
+        return ".".join(map(str, sys.version_info[:3]))
+
+    @staticmethod
+    def _get_package_version(package_name):
+        try:
+            return __import__(package_name).__version__
+        except (ImportError, AttributeError):
+            return None
+
+    @staticmethod
+    def get_default_build_dependencies():
+        build_dependencies = []
+        for package in ["pip", "setuptools", "wheel"]:
+            version = PythonEnv._get_package_version(package)
+            dep = (package + "==" + version) if version else package
+            build_dependencies.append(dep)
+        return build_dependencies
+
+    def to_dict(self):
+        return self.__dict__.copy()
+
+    @classmethod
+    def from_dict(cls, dct):
+        return cls(**dct)
+
+    def to_yaml(self, path):
+        with open(path, "w") as f:
+            # Exclude None and empty lists
+            data = {k: v for k, v in self.to_dict().items() if v}
+            yaml.safe_dump(data, f, sort_keys=False, default_flow_style=False)
+
+    @classmethod
+    def from_yaml(cls, path):
+        with open(path) as f:
+            return cls.from_dict(yaml.safe_load(f))
+
+    @classmethod
+    def from_conda_yaml(cls, path):
+        with open(path) as f:
+            conda_env = yaml.safe_load(f)
+
+        python = None
+        dependencies = None
+        for dep in conda_env.get("dependencies", []):
```

Comment:
Can the code here handle the case :
```
dependencies # model dependencies
-r requirements.txt
```
?

Reply:
```
dependencies # model dependencies
-r requirements.txt
```

Is this valid conda.yaml format?

Reply:
I haven't test. Maybe conda does not support it.

Reply:
> Maybe conda does not support it.

Please confirm first.

Reply:
@WeichenXu123 

```bash
% cat conda.yaml
name: fastai-example
channels:
  - conda-forge
dependencies: -r requirements.txt
                                                                                                                                                                 
% conda env create -f conda.yaml

CondaValueError: invalid package specification:  
```

Reply:
But do we handle:

```
dependencies:
    pip:
        -r requirements.txt
```

Reply:
@dbczumar I tried `conda env create` with that format. It didn't fail but skipped pip package installation.

```yaml
name: a
channels:
  - conda-forge
dependencies:
  pip: -r a.txt
```

```
% conda env create -f a.yaml
Collecting package metadata (repodata.json): done
Solving environment: done

Downloading and Extracting Packages
openssl-3.0.2        | 2.5 MB    | ########################################################################################################################### | 100% 
zlib-1.2.11          | 89 KB     | ########################################################################################################################### | 100% 
libzlib-1.2.11       | 60 KB     | ########################################################################################################################### | 100% 
setuptools-61.3.0    | 1.3 MB    | ########################################################################################################################### | 100% 
python-3.10.4        | 13.2 MB   | ########################################################################################################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate foo-bar-foo
#
# To deactivate an active environment, use
#
#     $ conda deactivate
```

---

```
name: b
channels:
  - conda-forge
dependencies:
  pip:
    - -r a.txt
```

```yaml
% conda env create -f a.yaml
Collecting package metadata (repodata.json): done
Solving environment: done

Downloading and Extracting Packages
openssl-3.0.2        | 2.5 MB    | ########################################################################################################################### | 100% 
zlib-1.2.11          | 89 KB     | ########################################################################################################################### | 100% 
libzlib-1.2.11       | 60 KB     | ########################################################################################################################### | 100% 
setuptools-61.3.0    | 1.3 MB    | ########################################################################################################################### | 100% 
python-3.10.4        | 13.2 MB   | ########################################################################################################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate foo-bar-foo
#
# To deactivate an active environment, use
#
#     $ conda deactivate
```

Reply:
Understood! Thanks for checking - nice to not have to handle this.